Initial attempt for a regex based generator

jreinhardt opened the pull request:
Hi,

this is an rough, initial attempt to create a strategy that builds strings that match a given python regex, with some inspiration form xeger (https://github.com/crdoconnor/xeger). It kind of works (negated ranges seem to be buggy), but it is extremely slow.

I am pretty sure that this is not the right way to implement this functionality, but I am willing to rewrite it in a better way, if such functionality is useful for hypothesis and if you could give me some pointers how to do it properly.

Best regards

DRMacIver commented:
Hi.

This is interesting work, thanks. I'm definitely not against including it if you can get it polished up and working well.

In terms of how to get it more suitable for Hypothesis, there are a bunch of stylistic things (Python 2 only code, print statements, etc) I assume you know about but I can talk in more detail about if you want. The first big thing is that it's probably going to need a lot of testing given how complicated the range of supported regex is.

In terms of performance, the big thing that's probably killing you right now is going to be strategy creation on every state transition. A thing that might help you do that is to create an explicit SearchStrategy subclass (using this isn't much different from composite really) so you can manage a bit more state and then cache the calculations so that each state only gets a strategy calculated for it once.

I don't have much more immediate feedback than that, but probably will as this progresses. Is there something specific you'd like me to comment on that I haven't?

jreinhardt commented:
I worked some more on this Pull Request. The code is now python 2/3 compatible, it's now a SearchStrategy and the performance is much better. I fixed some bugs and added a few more features, like lookahead assertions.

For tests there now is a corpus of regular expressions for which I test that generated strings match the regex. This corpus provides full coverage of the code.

It now supports every feature that xeger supports and some, and some bugs fixed. However not every python regex is supported, some of the non-regular features are missing like lookbehind assertions and conditional matching (?(id/name)yes-pattern|no-pattern).

What else is needed before you consider it ready?

DRMacIver commented on tests/cover/test_regex.py:
```diff
@@ -0,0 +1,123 @@
+])
+def test_re_examples(name, pattern):
+    """This test exercises all examples from the python documentation."""
+    find(st.regex(pattern), lambda x: re.match(pattern, x) is not None)
```
So currently this test is just asserting that you can find *one* value matching the pattern, when it should really be asserting that all generated values match the pattern.

I think this would be better with a nested given. Something like the following:

```python
@given(st.regex(pattern))
def test_all_generated_data_matches(x):
    assert re.match(pattern, x) is not None
test_all_generated_data_matches()
```

(you need to invoke nested givens explicitly because the test runner won't pick them up)

DRMacIver commented on src/hypothesis/searchstrategy/regex.py:
```diff
@@ -0,0 +1,150 @@
+
+    def __init__(self, pattern):
+        parsed = re.sre_parse.parse(pattern)
+        self.cache = {}
```
I don't think I entirely understand how you're using this cache. I was expecting all calls to handle\_state to be cached, but that seems mostly not to be the case?

Also, given that you're explicitly calculating the exact strategy you need and don't need it past that point, I wonder if it would make more sense to have a global state to strategy cache shared between all regex strategies.

jreinhardt replied:
This cache is for resolving group references, i.e. \1 or (?P<name>...) see l. 115 and l. 126.

Cacheing would indeed only make sense only on a global level. But I don't think I understand typical workloads well enough to have an opinion on its effectivity.

jreinhardt commented on src/hypothesis/searchstrategy/regex.py:
```diff
@@ -0,0 +1,150 @@
+import string
+from itertools import chain
+
+import hypothesis.strategies as strats
```
OK

jreinhardt commented on src/hypothesis/searchstrategy/regex.py:
```diff
@@ -0,0 +1,150 @@
+        self.strategies = [self._handle_state(state) for state in parsed]
+
+    def do_draw(self, data):
+        self.cache = {}
```
See above

jreinhardt commented on src/hypothesis/searchstrategy/regex.py:
```diff
@@ -0,0 +1,150 @@
+
+    def do_draw(self, data):
+        self.cache = {}
+        return u"".join(data.draw(strat) for strat in self.strategies)
```
No reason, just missed it.

jreinhardt commented on tests/cover/test_regex.py:
```diff
@@ -0,0 +1,123 @@
+    find(st.regex(pattern), lambda x: re.match(pattern, x) is not None)
+
+
+# Try to exhaust the set of matching strings for a number of simple res
```
It is lost, I did not want to commit that ...

jreinhardt commented on src/hypothesis/strategies.py:
```diff
@@ -1067,6 +1068,12 @@ def __not_a_first_class_strategy(self, name):
     return DataStrategy()
 
+
+@defines_strategy
```
ok

jreinhardt commented on src/hypothesis/strategies.py:
```diff
@@ -1067,6 +1068,12 @@ def __not_a_first_class_strategy(self, name):
 
+
+@defines_strategy
+def regex(pattern):
```
ok

jreinhardt commented on tests/cover/test_regex.py:
```diff
@@ -0,0 +1,123 @@
+    #    ("negativeLookbehind","
+    ('groupRef', '(a|b) \1')
+])
+def test_coverage(name, pattern):
```
the test below should not have been committed...

jreinhardt commented on src/hypothesis/searchstrategy/regex.py:
```diff
@@ -0,0 +1,150 @@
+        elif opcode == 'assert_not':
+            return strats.just('')
+        elif opcode == 'groupref':
+            return strats.shared(self.cache[value], key=value)
```
value is the index/name of the group that is referenced, i.e. is unique per pattern. The intention is that the string generated by the shared strategy is the same as the group it references in this draw. 

Maybe I misunderstood the purpose or scope of shared.

DRMacIver replied:
shared will draw a value from the provided strategy the first time and then reuse that value for all subsequent draws within this test run against the same key.

The problem here is that if I have two regular expressions which share a group this seems like it will end up sharing the values between them? I think I'm maybe still not clear on what you're trying to do in this bit.

DRMacIver replied:
Yeah, I don't think shared is a good way of doing this. it would cause a bunch of problems. e.g. if you had `lists(regex(r"<([a-z]+)>.*</\1>"))` then every regex in the list would have to have the same value for `\1`

jreinhardt replied:
What alternatives do I have? I guess I could write a custom strategy that does that kind of sharing, but is there simpler/easier/better way that I am missing?

DRMacIver replied:
The easiest thing is probably to store some useful temporary state on the strategy instance during do\_draw and clear it up when do\_draw finishes. It's not ideal but I can't think of a better one at the moment.

mulkieran commented:
It is sort of fun to use hypothesis to test hypothesis strategies. I think you could do this very well if you wrote a strategy to generate regular expressions. I did something similar, but not the same, for things know as D-Bus signatures: https://github.com/stratis-storage/into-dbus-python/blob/master/tests/test_signature_parser.py. It worked out nicely, and might be a helpful model.

DRMacIver commented:
BTW a thing that occurred to me might be an interesting solution to the groups problem: If you can figure out an easy way to transform the regex parse tree, the right thing to do is probably to extract all the groups, generate them up front and then use flatmap or similar to draw given those.

mlakewood commented:
@DRMacIver pointed me this way as I wanted something exactly like this. I can confirm the changes needed for matching the Regex opcodes in python3. Not sure how you are going to get around that.

Im also concerned about the speed at which this generates. I've had to suppress the HealthChecks for both `data_to_large` and `too_slow` when using this regex. As im completely inexperienced with this kind of thing im not sure how to help with this. So just adding as a data point.

However Im totally looking forward to having this in Hypothesis so really encourage this getting finished and merged.

Thanks!

jreinhardt commented:
I haven't yet looked into the details of the regex changes in 3.4, but I don't want to spend so a lot of time writing compatibility layers.

What do you think about buildiung not on the re module from the std library, but on an some other engine, like https://pypi.python.org/pypi/regex?

DRMacIver commented:
I don't in principle have a problem with building it on an external regex package that is syntax compatible with re. It would have to go into hypothesis.extra instead though (which is still part of the main Hypothesis package, but is where things with optional external dependencies go).

That being said I think you're overestimating the difficulty of building the compatibility layer and underestimating the difficulty of figuring out how to recreate all of this on top of a different API and engine.

lucaswiman commented on src/hypothesis/searchstrategy/regex.py:
```diff
@@ -0,0 +1,150 @@
+        elif category == 'category_not_space':
+            return string.printable.strip()
+        elif category == 'category_word':
+            return string.ascii_letters + string.digits + '_'
```
I'd guess this method is the source of a lot of the slowness, since these character sets are recomputed on each draw. It would be worth caching them, or just generating them on module import.

jreinhardt replied:
That is indeed true, and storing them on a static variable is actually much nicer.

lucaswiman commented on src/hypothesis/searchstrategy/regex.py:
```diff
@@ -0,0 +1,150 @@
+        elif category == 'category_space':
+            return string.whitespace
+        elif category == 'category_not_space':
+            return string.printable.strip()
```
This depends on `string.printable` happening to have all the whitespace at the end. While that's unlikely to change, it might be better to compute these character classes as the set of 0-127 characters matched by the appropriate regular expression, e.g.:
```
not_digit = re.compile('[^\d]')
category_not_digit = ''.join(filter(not_digit.match, string.printable))
```

jreinhardt commented on src/hypothesis/searchstrategy/regex.py:
```diff
@@ -0,0 +1,150 @@
+            return ''.join(set(string.printable)
+                           .difference(string.ascii_letters +
+                                       string.digits + '_'))
+        else:
```
There are none that I am aware of right now, but I consider it defensive to not to silently pass through when this happens at some point for whatever reasons (changes in re, bugs ...). Catching unhandled cases makes identifying the problem much more difficult. (I also consider this in line with your manifesto).

So I changed it into an AssertionError, this expresses  the my intent more accurately . Or shall I create a custom subclass of Hypothesis Error?

jreinhardt replied:
Actually, when I tried to commit this, I noticed that I run into problems with the cover tests, as these lines are not hit, which is somehow their intended behaviour.

So I can remove these catch-all cases, making check-coverage happy, but make the code less robust in the case of unexpected input. Or I have to cheat check-coverage somehow.

What do prefer?

Zac-HD replied:
Certainly don't sacrifice error-handling for coverage - you can use the comment  `# pragma: no cover` to tell coverage you don't expect to ever hit that branch.  See [here](https://github.com/HypothesisWorks/hypothesis-python/blob/master/src/hypothesis/internal/conjecture/utils.py#L108) for an example.

jreinhardt commented on src/hypothesis/searchstrategy/regex.py:
```diff
@@ -0,0 +1,150 @@
+        elif opcode == 'category':
+            return self._categories(value)
+        else:
+            raise NotImplementedError
```
Same reply as above :)

jreinhardt commented on src/hypothesis/searchstrategy/regex.py:
```diff
@@ -0,0 +1,150 @@
+                result.append(part)
+            return strats.tuples(*result).map(lambda x: u"".join(x))
+        else:
+            raise NotImplementedError(opcode)
```
see above

jreinhardt commented:
Sorry for the long delay, I just started to work on it again.

Zac-HD commented:
The `xeger` library linked above seems to have deliberately dropped Python 3 compatibility; [the original source](https://bitbucket.org/leapfrogdevelopment/rstr/src/default/rstr/xeger.py#xeger.py-103:107) has a small function to convert opcode strings or constants into a consistent form.

Zac-HD commented on src/hypothesis/searchstrategy/regex.py:
```diff
@@ -0,0 +1,154 @@
+def strategy_concat(strategies):
+    """given a list of strategies yielding strings return a strategy yielding
+    their concatenation."""
+    return st.tuples(*strategies).map(lambda x: u"".join(x))
```
This could be `.map(u''.join)` - no need for a lambda when it's already a one-argument callable.

Zac-HD commented on src/hypothesis/searchstrategy/regex.py:
```diff
@@ -0,0 +1,154 @@
+                    max_size=end_range
+                ).map(lambda x: u"".join(x))
+                result.append(part)
+            return st.tuples(*result).map(lambda x: u"".join(x))
```
Or `strategy_concat(result)` here.

mulkieran commented on src/hypothesis/searchstrategy/regex.py:
```diff
@@ -0,0 +1,154 @@
+                    self._handle_state(v),
+                    min_size=start_range,
+                    max_size=end_range
+                ).map(lambda x: u"".join(x))
```
Same opportunity as suggested above:

```lambda x: u"".join(x)``` -> ```u"".join```

Zac-HD commented:
I'm closing this as superseded by #776, because we have enough open pulls that it's starting to get confusing.  @jreinhardt, thanks for your contribution and I'm sorry that we weren't able to get it merged - regex is a lot harder than I expected 😞
